core/persistence.py

```diff
@@ -141,47 +142,93 @@ async def delete_blocked_id(self, blocked_id: int) -> None:
+        """
+        return tuple(
+            (row["query"], row["guild_id"])
+            for row in await self.execute_fetchall(
+                (
+                    f"SELECT query, guild_id FROM highlight WHERE user_id = ? {'AND guild_id IN (?, 0)' if guild_id else ''}"
+                    f"{' ORDER BY guild_id' if order_by_guild else ''}"
+                ),
+                (user_id, guild_id) if guild_id else (user_id,),
+            )
+        )
+
+    async def fetch_user_highlight_guilds(self, user_id: int, pattern: str, exclude_global_guild=False) -> tuple[int]:
+        return tuple(
+            row["guild_id"]
+            for row in await self.execute_fetchall(
+                (
+                    "SELECT guild_id FROM highlight WHERE user_id = ? AND query = ?"
+                    f"{' AND guild_id != 0' if exclude_global_guild else ''}"
+                ),
+                (user_id, pattern),
+            )
+        )
 
     async def fetch_user_highlight_count(self, user_id: int) -> int:
         return (await self.execute_fetchone("SELECT COUNT(query) FROM highlight WHERE user_id = ?", (user_id,)))[0]  # type: ignore
 
-    async def fetch_user_highlights_starting_with(self, user_id: int, prefix: str) -> Tuple[str, ...]:
-        rows = await self.execute_fetchall(
-            "SELECT query FROM highlight WHERE user_id = ? AND query LIKE ?", (user_id, f"{prefix}%")
-        )
-        return tuple(row[0] for row in rows)
-
-    async def delete_highlight(self, user_id: int, query: str) -> None:
-        await self.execute("DELETE FROM highlight WHERE user_id = ? AND query = ?", (user_id, query))
+    async def fetch_user_highlights_starting_with(
+        self, user_id: int, prefix: str, guild_id: int = 0
+    ) -> tuple[tuple[str, int], ...]:
+        if guild_id:
+            rows = await self.execute_fetchall(
+                "SELECT query, guild_id FROM highlight WHERE user_id = ? AND (guild_id = ? OR guild_id = 0) AND query LIKE ?",
+                (user_id, guild_id, f"{prefix}%"),
+            )
+        else:
+            rows = await self.execute_fetchall(
+                "SELECT query, guild_id FROM highlight WHERE user_id = ? AND query LIKE ?", (user_id, f"{prefix}%")
+            )
+        return tuple(tuple(row) for row in rows)
+
+    async def delete_highlight(self, user_id: int, query: str, guild_id: int = 0) -> None:
+        if guild_id:
+            await self.execute(
+                "DELETE FROM highlight WHERE user_id = ? AND guild_id = ? AND query = ?", (user_id, guild_id, query)
+            )
+        else:
+            await self.execute("DELETE FROM highlight WHERE user_id = ? AND query = ?", (user_id, query))
 
-    async def clear_user_highlights(self, user_id: int) -> None:
-        await self.execute("DELETE FROM highlight WHERE user_id = ?", (user_id,))
+    async def clear_user_highlights(self, user_id: int, guild_id: int = 0) -> None:
+        if guild_id:
```

Comment:
Ok, good, I see you prefer this explicitly in general, will keep it in mind.

Just a thought:
With restricted types / type checking it makes no difference, besides arguably readability. (Might type checking via mypy e.g. then be a good option in the workflow to enforce it?)
Without strict typing (since it is technically optional in python), just `if guild_id` would catch e.g. passed `None` values, instead of passing `NULL` to the query, resulting in deleting nothing (or returning nothing above).
But I understand your point.

Reply:
I use Pyright for type checking, but type safety is not enforced in the CI since there are some places where types aren't correctly recognised (for example when converters are used as type hints instead of the actual type). The explicit comparisons are just a matter of taste. In my opinion, it's more readable and easier to understand, especially if you're coming from a different programming language.